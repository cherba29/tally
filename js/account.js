goog.provide('budget.Account');

goog.require('budget.Month');

goog.scope(function() {

/**
 * @final
 * @constructor
 */
budget.Account = function() {

};
var Account = budget.Account;

/**
 * @type {string}
 * @private
 */
Account.name_ = null;

/**
 * @type {string}
 * @private
 */
Account.description_ = null;

/**
 * @type {string}
 * @private
 */
Account.number_ = null;
Account.owners_ = [];

/**
 * @type {Month}
 * @private
 */
Account.openedOn_ = null;

/**
 * @type {Month}
 * @private
 */
Account.closedOn_ = null;

/**
 * @type {string}
 * @private
 */
Account.url_ = null;

/**
 * @type {string}
 * @private
 */
Account.userName_ = null;

/**
 * @type {string}
 * @private
 */
Account.password_ = null;

/**
 * @type {string}
 * @private
 */
Account.phone_ = null;

/**
 * @type {string}
 * @private
 */
Account.address_ = null;

});  // goog.scope

goog.exportSymbol('budget.Account', budget.Account);
