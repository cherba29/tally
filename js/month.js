goog.provide('budget.Month');

goog.scope(function() {

/**
 * @param {number=} year Year for the month.
 * @param {number=} month Month [0-11] for this month instance. 
 * @final
 * @constructor
 */
budget.Month = function(year, month) {

};
var Account = budget.Account;

/**
 * @type {string}
 * @private
 */
Account.name_ = null;

/**
 * @type {string}
 * @private
 */
Account.description_ = null;

/**
 * @type {string}
 * @private
 */
Account.number_ = null;
Account.owners_ = [];

/**
 * @type {Month}
 * @private
 */
Account.openedOn_ = null;

/**
 * @type {Month}
 * @private
 */
Account.closedOn_ = null;

/**
 * @type {string}
 * @private
 */
Account.url_ = null;

/**
 * @type {string}
 * @private
 */
Account.userName_ = null;

/**
 * @type {string}
 * @private
 */
Account.password_ = null;

/**
 * @type {string}
 * @private
 */
Account.phone_ = null;

/**
 * @type {string}
 * @private
 */
Account.address_ = null;

});  // goog.scope

goog.exportSymbol('budget.Account', budget.Account);
