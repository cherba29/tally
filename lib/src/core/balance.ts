export enum Type {
  UNKNOWN = 'UNKNOWN',
  CONFIRMED = 'CONFIRMED',
  PROJECTED = 'PROJECTED',
  AUTO_PROJECTED = 'AUTO_PROJECTED'
}

const combineTypes = (t1: Type, t2: Type): Type => (t1 < t2 ? t2 : t1);

export class Balance {
  constructor(readonly amount: number, readonly date: Date, readonly type: Type) {}

  // Helper contructor.
  static confirmed(amount: number, date: string): Balance {
    return new Balance(amount, new Date(date), Type.CONFIRMED);
  }

  static projected(amount: number, date: string): Balance {
    return new Balance(amount, new Date(date), Type.PROJECTED);
  }

  static negated(balance: Balance): Balance {
    return new Balance(-balance.amount, balance.date, balance.type);
  }

  static add(balance1: Balance, balance2: Balance): Balance {
    const maxDate = balance1.date.getTime() < balance2.date.getTime() ? balance2.date : balance1.date;
    return new Balance(
      balance1.amount + balance2.amount,
      maxDate,
      combineTypes(balance1.type, balance2.type)
    );
  }

  static subtract(balance1: Balance, balance2: Balance): Balance {
    return Balance.add(balance1, new Balance(-balance2.amount, balance2.date, balance2.type));
  }

  compareTo(other: Balance): number {
    const dateDiff = this.date.getTime() - other.date.getTime();
    if (dateDiff !== 0) {
      return dateDiff;
    }
    const amountDiff = this.amount - other.amount;
    if (amountDiff !== 0) {
      return amountDiff;
    }
    return this.type.localeCompare(other.type);
  }

  toString(): string {
    return `Balance { amount: ${this.amount}, date: ${this.date
      .toISOString()
      .slice(0, 10)}, type: ${Type[this.type]} }`;
  }
}
