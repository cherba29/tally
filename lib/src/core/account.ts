import { Month } from './month';

export enum Type {
  UNSPECIFIED = '_unspecified_',
  BILL = 'bill',
  CHECKING = 'checking',
  CREDIT = 'credit',
  CREDIT_CARD = 'credit-card',
  DEFERRED_INCOME = 'deferred income',
  EXTERNAL = 'external',
  INCOME = 'income',
  INVESTMENT = 'investment',
  RETIREMENT = 'retirement',
  SUMMARY = '_summary_',
  TAX = 'tax_',
}

// Typescript cannot reverse map string enums, so prebuild a lookup map.
// https://www.typescriptlang.org/docs/handbook/enums.html
const reverseAccountType = new Map<string, string>(Object.entries(Type).map(([k, v]) => [v, k]));

export interface InitData {
  name: string;
  description?: string;
  parents?: Account[];
  children?: Account[];
  autoBalance?: boolean;
  type: Type;
  number?: string;
  openedOn?: Month;
  closedOn?: Month;
  owners: string[];
  url?: string;
  address?: string;
  phone?: string;
  userName?: string;
  password?: string;
}

export class Account {
  readonly name: string;
  readonly description?: string;
  readonly parents: Account[];
  readonly children: Account[];
  // Account type, for example 'CREDIT_CARD'.
  readonly type: Type;
  // Account balances are automatically considered to be committed.
  readonly autoBalance: boolean;
  // Real account number associated with this account.
  readonly number?: string;
  // Month when account was opened and possibly closed.
  readonly openedOn?: Month;
  readonly closedOn?: Month;
  // List of account owner ids.
  readonly owners: string[];
  // Url to the account.
  readonly url?: string;
  // Physical address for the account.
  readonly address?: string;
  // Phone number for customer support.
  readonly phone?: string;
  // Username/password to use to login to the account.
  readonly userName?: string;
  readonly password?: string;

  constructor(data: InitData) {
    this.name = data.name;
    this.description = data.description;
    this.parents = data.parents ?? [];
    this.children = data.children ?? [];
    this.autoBalance = data.autoBalance ?? false;
    this.type = data.type;
    this.number = data.number;
    this.openedOn = data.openedOn;
    this.closedOn = data.closedOn;
    this.owners = data.owners;
    this.url = data.url;
    this.address = data.address;
    this.phone = data.phone;
    this.userName = data.userName;
    this.password = data.password;
  }

  toString() {
    return `Account ${this.name} ${this.type}${this.closedOn ? ` Closed ${this.closedOn}` : ''}`;
  }

  isClosed(month: Month): boolean {
    return !!(
      (
        this.closedOn?.isLess(month) || // After closed.
        (this.openedOn && month.isLess(this.openedOn))
      ) // Before or on open.
    );
  }

  // Is this an external/bookeeping account. External accounts are not
  // considered to be part of any owner.
  get isExternal(): boolean {
    return (
      this.type === Type.EXTERNAL || this.type === Type.TAX || this.type === Type.DEFERRED_INCOME
    );
  }

  // Is this synthetic summary account.
  get isSummary(): boolean {
    return this.type === Type.SUMMARY;
  }

  hasCommonOwner(other: Account): boolean {
    return this.owners.some((owner) => other.owners.includes(owner));
  }

  get typeIdName(): string {
    return reverseAccountType.get(this.type.toString()) || '';
  }
}
