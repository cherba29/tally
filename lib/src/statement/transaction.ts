import { Account } from '@tally/lib/core/account';
import { Balance, Type as BalanceType } from '@tally/lib/core/balance';
import { Budget } from '@tally/lib/core/budget';
import { Month } from '@tally/lib/core/month';
import { Transfer } from '@tally/lib/core/transfer';
import { Statement } from './statement';

export enum Type {
  UNKNOWN,
  TRANSFER,
  INCOME,
  EXPENSE
}

export interface Transaction {
  account: Account;
  balance: Balance;
  description?: string;
  type: Type;
  balanceFromStart?: number;
  balanceFromEnd?: number;
}

// Extension of Statement for transactions over an account.
export class TransactionStatement extends Statement {
  // Account to which this transaction statements belongs.
  account: Account;

  // List of transcation in this statement.
  transactions: Transaction[] = [];

  // True if any transactions in this statement "cover" previous statement.
  coversPrevious = false;

  // True if any projected transactions in this statement "cover"
  // previous statement.
  coversProjectedPrevious = false;

  // True if any of the transcations are projects.
  hasProjectedTransfer = false;

  // True if this statement is covered by next.
  isCovered = false;

  // True if this statement is covered by any projected transactions in next
  // statement.
  isProjectedCovered = false;

  // Is this statement for closed account for given month.
  get isClosed(): boolean {
    return this.account.isClosed(this.month);
  }

  constructor(account: Account, month: Month) {
    super(account.name, month);
    this.account = account;
  }
}

function getTransactionType(fromAccount: Account, toAccount: Account, amount: number): Type {
  if (toAccount.hasCommonOwner(fromAccount) && !toAccount.isExternal && !fromAccount.isExternal) {
    return Type.TRANSFER;
  } else {
    return amount > 0 ? Type.INCOME : Type.EXPENSE;
  }
}

function makeTranscationStatement(
  account: Account,
  month: Month,
  transfers?: Set<Transfer>,
  startBalance?: Balance
): TransactionStatement {
  const statement = new TransactionStatement(account, month);
  statement.startBalance = startBalance;
  const attributeTransfer = (fromAccount: Account, toAccount: Account, amount: number): Type => {
    if (amount > 0) {
      statement.inFlows += amount;
    } else {
      statement.outFlows += amount;
    }
    const transactionType = getTransactionType(fromAccount, toAccount, amount);
    switch (transactionType) {
      case Type.EXPENSE:
        statement.totalPayments += amount;
        break;
      case Type.INCOME:
        statement.income += amount;
        break;
      case Type.UNKNOWN:
        break;
      case Type.TRANSFER:
        statement.totalTransfers += amount;
        break;
    }
    return transactionType;
  };
  const descTransfers = Array.from(transfers || []).sort((b, a) => {
    let eq = a.balance.compareTo(b.balance);
    if (eq != 0) return eq;
    eq = a.fromMonth.compareTo(b.fromMonth);
    if (eq != 0) return eq;
    eq = a.toMonth.compareTo(b.toMonth);
    if (eq != 0) return eq;
    if (a.fromAccount.name != b.fromAccount.name) {
      return a.fromAccount.name < b.fromAccount.name ? -1 : 1;
    }
    if (a.toAccount.name != b.toAccount.name) {
      return a.toAccount.name < b.toAccount.name ? -1 : 1;
    }
    if (a.description != b.description) {
      return (a.description || '') < (b.description || '') ? -1 : 1;
    }
    return 0;
  });

  const firstTransfer: Transfer|undefined = descTransfers[descTransfers.length - 1];
  if (firstTransfer && startBalance && firstTransfer.balance.date.getTime() < startBalance.date.getTime()) {
    throw new Error(
      `Balance ${month} ${startBalance} for account ${account.name} ` + 
      `starts after transaction ${firstTransfer.fromAccount.name} --> ${firstTransfer.toAccount.name}/${firstTransfer.balance} ` +
      `desc "${firstTransfer.description}"`);
  }

  for (const t of descTransfers) {
    statement.hasProjectedTransfer =
      statement.hasProjectedTransfer || t.balance.type == BalanceType.PROJECTED;
    let otherAccount: Account | undefined;
    let balance: Balance | undefined;
    let transactionType = Type.UNKNOWN;
    if (t.toAccount.name === account.name) {
      balance = t.balance;
      otherAccount = t.fromAccount;
      transactionType = attributeTransfer(otherAccount, account, balance.amount);
    } else if (t.fromAccount.name === account.name) {
      balance = Balance.negated(t.balance);
      otherAccount = t.toAccount;
      transactionType = attributeTransfer(account, otherAccount, balance.amount);
    } else {
      // This should never occur since budget should have been validated by now.
      throw new Error(
        `Setting transfer (${t.fromAccount.name} to ${t.toAccount.name}) for ${account.name} account statement!`
      );
    }
    if (!statement.coversPrevious && balance.amount > 0 && t.fromAccount.hasCommonOwner(account)) {
      statement.coversProjectedPrevious = true;
      if (balance.type != BalanceType.PROJECTED) {
        statement.coversPrevious = true;
      }
    }
    statement.transactions.push({
      account: otherAccount,
      description: t.description,
      balance,
      type: transactionType
    });
  }
  return statement;
}

export function buildTransactionStatementTable(budget: Budget): TransactionStatement[] {
  const statementTable: TransactionStatement[] = [];

  const makeStatement = (account: Account, month: Month): TransactionStatement => {
    return makeTranscationStatement(
      account,
      month,
      budget.transfers.get(account.name)?.get(month.toString()),
      budget.balances.get(account.name)?.get(month.toString())
    );
  };

  // Working backwards.
  const months = budget.months.concat().sort((a, b) => b.compareTo(a));
  if (months.length < 1) {
    throw new Error('Budget must have at least one month.');
  }

  for (const account of budget.accounts.values()) {
    const accountStatements: TransactionStatement[] = [];
    // Make statement outside range so that its attributes relating to previous can be used.
    let nextMonthStatement = makeStatement(account, months[0]!.next());
    for (const month of months) {
      const statement = makeStatement(account, month);
      statement.endBalance = nextMonthStatement.startBalance;
      statement.isCovered =
        statement.endBalance === undefined ||
        statement.endBalance.amount >= 0 ||
        nextMonthStatement.coversPrevious;
      statement.isProjectedCovered =
        statement.isCovered || nextMonthStatement.coversProjectedPrevious;
      nextMonthStatement = statement;
      if (statement.startBalance) {
        statement.transactions.reduceRight((prevBalance, t) => {
          return (t.balanceFromStart = prevBalance + t.balance.amount);
        }, statement.startBalance.amount);
      }
      if (statement.endBalance) {
        statement.transactions.reduce((prevBalance, t) => {
          return (t.balanceFromEnd = prevBalance - t.balance.amount);
        }, statement.endBalance.amount);
      }
      accountStatements.push(statement);
    }
    // Do not include account if for all months it was closed.
    if (accountStatements.some((stmt) => !stmt.isClosed)) {
      statementTable.push(...accountStatements);
    }
  }
  return statementTable;
}
