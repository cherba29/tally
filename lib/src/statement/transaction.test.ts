import { describe, expect, test } from '@jest/globals';
import { Account, Type as AccountType } from '../core/account';
import { Balance } from '../core/balance';
import { BudgetBuilder } from '../core/budget';
import { Month } from '../core/month';
import { buildTransactionStatementTable, Type as TransactionType } from './transaction';

describe('Build', () => {
  test('no months', () => {
    const builder = new BudgetBuilder();
    expect(() => buildTransactionStatementTable(builder.build())).toThrow(
      new Error('Budget must have at least one month.')
    );
  });

  test('single account no transfers', () => {
    const builder = new BudgetBuilder();
    const account = new Account({
      name: 'test-account',
      type: AccountType.EXTERNAL,
      owners: [],
      openedOn: new Month(2019, 11),
    });
    builder.setAccount(account);
    const table = buildTransactionStatementTable(builder.build());
    expect(table).toEqual([
      {
        account,
        coversPrevious: false,
        coversProjectedPrevious: false,
        endBalance: undefined,
        hasProjectedTransfer: false,
        inFlows: 0,
        income: 0,
        isCovered: true,
        isProjectedCovered: true,
        month: new Month(2019, 11),
        name: 'test-account',
        outFlows: 0,
        startBalance: undefined,
        totalPayments: 0,
        totalTransfers: 0,
        transactions: [],
      },
    ]);
    expect(table[0]?.isClosed).toBe(false);
  });

  test('bad account name on transfer', () => {
    const builder = new BudgetBuilder();
    const account1 = new Account({
      name: 'test-account1',
      type: AccountType.CHECKING,
      owners: ['john'],
    });
    builder.setAccount(account1);
    builder.addTransfer({
      fromAccount: 'test-account1',
      toAccount: 'test-account2',
      toMonth: Month.fromString('Dec2019'),
      fromMonth: Month.fromString('Dec2019'),
      balance: Balance.projected(2000, '2019-12-05'),
      description: 'First transfer',
    });
    expect(() => buildTransactionStatementTable(builder.build())).toThrow(
      new Error('Unknown account test-account2')
    );
  });

  test('two accounts with common owner and transfers', () => {
    const builder = new BudgetBuilder();
    const account1 = new Account({
      name: 'test-account1',
      type: AccountType.CHECKING,
      owners: ['john'],
    });
    const account2 = new Account({
      name: 'test-account2',
      type: AccountType.CREDIT,
      owners: ['john'],
    });

    builder.setAccount(account1);
    builder.setBalance('test-account1', 'Dec2019', Balance.confirmed(10, '2019-12-01'));
    builder.setBalance('test-account1', 'Jan2020', Balance.confirmed(20, '2020-01-01'));
    builder.setBalance('test-account1', 'Feb2020', Balance.projected(30, '2020-02-01'));
    builder.setAccount(account2);
    builder.addTransfer({
      fromAccount: 'test-account1',
      toAccount: 'test-account2',
      toMonth: Month.fromString('Dec2019'),
      fromMonth: Month.fromString('Dec2019'),
      balance: Balance.projected(2000, '2019-12-05'),
      description: 'First transfer',
    });

    builder.addTransfer({
      fromAccount: 'test-account1',
      toAccount: 'test-account2',
      toMonth: Month.fromString('Dec2019'),
      fromMonth: Month.fromString('Dec2019'),
      balance: Balance.projected(1000, '2019-12-05'),
      description: 'Second transfer',
    });

    const table = buildTransactionStatementTable(builder.build());
    expect(table).toMatchSnapshot();
  });

  test('two accounts with extenal transfer', () => {
    const builder = new BudgetBuilder();
    const account1 = new Account({
      name: 'test-account1',
      type: AccountType.EXTERNAL,
      owners: [],
    });
    const account2 = new Account({
      name: 'test-account2',
      type: AccountType.CREDIT,
      owners: ['john'],
    });

    builder.setAccount(account1);
    builder.setBalance('test-account1', 'Dec2019', Balance.confirmed(10, '2019-12-01'));
    builder.setBalance('test-account1', 'Jan2020', Balance.confirmed(20, '2020-01-01'));
    builder.setBalance('test-account1', 'Feb2020', Balance.projected(30, '2020-02-01'));
    builder.setAccount(account2);
    builder.addTransfer({
      fromAccount: 'test-account1',
      toAccount: 'test-account2',
      toMonth: Month.fromString('Dec2019'),
      fromMonth: Month.fromString('Dec2019'),
      balance: Balance.projected(2000, '2019-12-05'),
      description: 'First transfer',
    });

    builder.addTransfer({
      fromAccount: 'test-account1',
      toAccount: 'test-account2',
      toMonth: Month.fromString('Dec2019'),
      fromMonth: Month.fromString('Dec2019'),
      balance: Balance.projected(1000, '2019-12-05'),
      description: 'Second transfer',
    });

    const table = buildTransactionStatementTable(builder.build());
    expect(table).toMatchSnapshot();
  });
});
