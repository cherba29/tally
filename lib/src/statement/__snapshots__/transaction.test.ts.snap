// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build two accounts with common owner and transfers 1`] = `
[
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "checking",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 1,
      "year": 2020,
    },
    "name": "test-account1",
    "outFlows": 0,
    "startBalance": Balance {
      "amount": 30,
      "date": 2020-02-01T00:00:00.000Z,
      "type": "PROJECTED",
    },
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "checking",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": Balance {
      "amount": 30,
      "date": 2020-02-01T00:00:00.000Z,
      "type": "PROJECTED",
    },
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 0,
      "year": 2020,
    },
    "name": "test-account1",
    "outFlows": 0,
    "startBalance": Balance {
      "amount": 20,
      "date": 2020-01-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "checking",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": Balance {
      "amount": 20,
      "date": 2020-01-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "hasProjectedTransfer": true,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 11,
      "year": 2019,
    },
    "name": "test-account1",
    "outFlows": -3000,
    "startBalance": Balance {
      "amount": 10,
      "date": 2019-12-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "totalPayments": 0,
    "totalTransfers": -3000,
    "transactions": [
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account2",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "credit",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": -2000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "balanceFromStart": -2990,
        "description": "First transfer",
        "type": 1,
      },
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account2",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "credit",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": -1000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "balanceFromStart": -990,
        "description": "Second transfer",
        "type": 1,
      },
    ],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 1,
      "year": 2020,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 0,
      "year": 2020,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": true,
    "endBalance": undefined,
    "hasProjectedTransfer": true,
    "inFlows": 3000,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 11,
      "year": 2019,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 3000,
    "transactions": [
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account1",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "checking",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": 2000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "description": "First transfer",
        "type": 1,
      },
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account1",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "checking",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": 1000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "description": "Second transfer",
        "type": 1,
      },
    ],
  },
]
`;

exports[`Build two accounts with extenal transfer 1`] = `
[
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [],
      "password": undefined,
      "phone": undefined,
      "type": "external",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 1,
      "year": 2020,
    },
    "name": "test-account1",
    "outFlows": 0,
    "startBalance": Balance {
      "amount": 30,
      "date": 2020-02-01T00:00:00.000Z,
      "type": "PROJECTED",
    },
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [],
      "password": undefined,
      "phone": undefined,
      "type": "external",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": Balance {
      "amount": 30,
      "date": 2020-02-01T00:00:00.000Z,
      "type": "PROJECTED",
    },
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 0,
      "year": 2020,
    },
    "name": "test-account1",
    "outFlows": 0,
    "startBalance": Balance {
      "amount": 20,
      "date": 2020-01-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account1",
      "number": undefined,
      "openedOn": undefined,
      "owners": [],
      "password": undefined,
      "phone": undefined,
      "type": "external",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": Balance {
      "amount": 20,
      "date": 2020-01-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "hasProjectedTransfer": true,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 11,
      "year": 2019,
    },
    "name": "test-account1",
    "outFlows": -3000,
    "startBalance": Balance {
      "amount": 10,
      "date": 2019-12-01T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "totalPayments": -3000,
    "totalTransfers": 0,
    "transactions": [
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account2",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "credit",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": -2000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "balanceFromStart": -2990,
        "description": "First transfer",
        "type": 3,
      },
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account2",
          "number": undefined,
          "openedOn": undefined,
          "owners": [
            "john",
          ],
          "password": undefined,
          "phone": undefined,
          "type": "credit",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": -1000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "balanceFromStart": -990,
        "description": "Second transfer",
        "type": 3,
      },
    ],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 1,
      "year": 2020,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": false,
    "inFlows": 0,
    "income": 0,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 0,
      "year": 2020,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [],
  },
  TransactionStatement {
    "account": Account {
      "address": undefined,
      "closedOn": undefined,
      "description": undefined,
      "name": "test-account2",
      "number": undefined,
      "openedOn": undefined,
      "owners": [
        "john",
      ],
      "password": undefined,
      "phone": undefined,
      "type": "credit",
      "url": undefined,
      "userName": undefined,
    },
    "coversPrevious": false,
    "coversProjectedPrevious": false,
    "endBalance": undefined,
    "hasProjectedTransfer": true,
    "inFlows": 3000,
    "income": 3000,
    "isCovered": true,
    "isProjectedCovered": true,
    "month": Month {
      "month": 11,
      "year": 2019,
    },
    "name": "test-account2",
    "outFlows": 0,
    "startBalance": undefined,
    "totalPayments": 0,
    "totalTransfers": 0,
    "transactions": [
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account1",
          "number": undefined,
          "openedOn": undefined,
          "owners": [],
          "password": undefined,
          "phone": undefined,
          "type": "external",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": 2000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "description": "First transfer",
        "type": 2,
      },
      {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account1",
          "number": undefined,
          "openedOn": undefined,
          "owners": [],
          "password": undefined,
          "phone": undefined,
          "type": "external",
          "url": undefined,
          "userName": undefined,
        },
        "balance": Balance {
          "amount": 1000,
          "date": 2019-12-05T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "description": "Second transfer",
        "type": 2,
      },
    ],
  },
]
`;
